code/datums/components/contraband.dm

```diff
@@ -0,0 +1,86 @@
+TYPEINFO(/datum/component/contraband)
+	initialization_args = list(
+		ARG_INFO("contraband_level", DATA_INPUT_NUM, "Non-firearm Contraband level this item will have", 0),
+		ARG_INFO("carry_level", DATA_INPUT_NUM, "Firearm Contraband level this item will have", 0),
+	)
+
+/datum/component/contraband
+	var/contraband_level = 0
+	var/carry_level = 0
+
+/datum/component/contraband/Initialize(contraband_level = 0, carry_level = 0)
+	. = ..()
+	if(!ismovable(parent))
+		return COMPONENT_INCOMPATIBLE
+	var/atom/movable/AM = parent
+	RegisterSignal(AM, COMSIG_MOVABLE_GET_CONTRABAND, PROC_REF(get_contraband))
+	SEND_SIGNAL(AM, COMSIG_MOVABLE_CONTRABAND_CHANGED)
+	src.contraband_level = contraband_level
+	src.carry_level = carry_level
+	RegisterSignal(parent, COMSIG_MOVABLE_CONTRABAND_CHANGED, PROC_REF(visible_contraband_changed))
+	if (isitem(AM))
+		var/obj/item/I = AM
+		if (ismob(I.loc))
+			var/mob/M = I.loc
+			src.equipped(I, M, I.equipped_in_slot)
+		RegisterSignal(AM, COMSIG_ITEM_EQUIPPED, PROC_REF(equipped))
+		RegisterSignal(AM, COMSIG_ITEM_PICKUP, PROC_REF(picked_up))
+		RegisterSignals(AM, list(COMSIG_ITEM_UNEQUIPPED, COMSIG_ITEM_DROPPED), PROC_REF(removed))
+	src.visible_contraband_changed(AM)
+
+/datum/component/contraband/proc/get_contraband(atom/owner, nonfirearms = TRUE, firearms = TRUE)
+	var/return_val = 0
+	if (nonfirearms && firearms)
+		return_val += HAS_ATOM_PROPERTY(owner, PROP_MOVABLE_CONTRABAND_OVERRIDE) ? GET_ATOM_PROPERTY(owner, PROP_MOVABLE_CONTRABAND_OVERRIDE) : (src.contraband_level + src.carry_level)
+		return_val += GET_ATOM_PROPERTY(owner, PROP_MOVABLE_VISIBLE_CONTRABAND)
+		return_val += GET_ATOM_PROPERTY(owner, PROP_MOVABLE_VISIBLE_GUNS)
+		return return_val
+	if (nonfirearms)
+		return_val += HAS_ATOM_PROPERTY(owner, PROP_MOVABLE_CONTRABAND_OVERRIDE) ? GET_ATOM_PROPERTY(owner, PROP_MOVABLE_CONTRABAND_OVERRIDE) : src.contraband_level
+		return_val += GET_ATOM_PROPERTY(owner, PROP_MOVABLE_VISIBLE_CONTRABAND)
+	if (firearms)
+		if (src.carry_level)
+			return_val += HAS_ATOM_PROPERTY(owner, PROP_MOVABLE_CONTRABAND_OVERRIDE) ? GET_ATOM_PROPERTY(owner, PROP_MOVABLE_CONTRABAND_OVERRIDE) : src.carry_level
+		return_val += GET_ATOM_PROPERTY(owner, PROP_MOVABLE_VISIBLE_GUNS)
+	return return_val
```

Comment:
this is an incorrect use of signal return values - the rule for signal returns is that they get bitwise ORed together, so any values you return from a proc invoked by a signal should be in form of bitflags (incl simple boolean TRUE/FALSE)

Can look at how the power cell and cell holder components handle the get_charge signal as an example of how to work around that restriction

Yes it's awkward for something like this where you're wanting to just return a number, but I think that maintaining consistency is worth the added code hassle